internal/cache/file/cache_handler_test.go

```diff
@@ -177,161 +176,239 @@ func (chrT *cacheHandlerTest) Test_createLocalFileReadHandle_OnlyForRead() {
+	existingJob := chrT.jobManager.GetJob(minObject.Name, chrT.bucket.Name())
+	AssertEq(nil, existingJob)
 
 	// Insertion will happen and that leads to eviction.
-	err := chrT.cacheHandler.addFileInfoEntryToCache(minObject, chrT.bucket)
+	err := chrT.cacheHandler.addFileInfoEntryAndCreateDownloadJob(minObject, chrT.bucket)
 
 	ExpectEq(nil, err)
 	ExpectTrue(chrT.isEntryInFileInfoCache(minObject.Name, chrT.bucket.Name()))
-	jobStatus := cacheHandle1.fileDownloadJob.GetStatus()
-	ExpectEq(downloader.INVALID, jobStatus.Name)
+	jobStatus := oldJob.GetStatus()
+	ExpectEq(downloader.Invalid, jobStatus.Name)
 	ExpectEq(false, doesFileExist(chrT.downloadPath))
+	// Job should be added for minObject
+	minObjectJob := chrT.jobManager.GetJob(minObject.Name, chrT.bucket.Name())
+	ExpectNe(nil, minObjectJob)
+	ExpectEq(downloader.NotStarted, minObjectJob.GetStatus().Name)
 }
 
-func (chrT *cacheHandlerTest) Test_addFileInfoEntryToCache_IfLocalFileGetsDeleted() {
-	// Existing cacheHandle.
-	cacheHandle1 := chrT.getCacheHandleForSetupEntryInCache()
-	AssertEq(nil, cacheHandle1.validateCacheHandle())
+func (chrT *cacheHandlerTest) Test_addFileInfoEntryAndCreateDownloadJob_IfLocalFileGetsDeleted() {
 	// Delete the local cache file.
 	err := os.Remove(chrT.downloadPath)
 	AssertEq(nil, err)
 
 	// There is a fileInfoEntry in the fileInfoCache but the corresponding local file doesn't exist.
 	// Hence, this will return error containing util.FileNotPresentInCacheErrMsg.
-	err = chrT.cacheHandler.addFileInfoEntryToCache(chrT.object, chrT.bucket)
+	err = chrT.cacheHandler.addFileInfoEntryAndCreateDownloadJob(chrT.object, chrT.bucket)
 
 	AssertNe(nil, err)
 	ExpectTrue(strings.Contains(err.Error(), util.FileNotPresentInCacheErrMsg))
 }
 
-func (chrT *cacheHandlerTest) Test_GetCacheHandle_WhenCacheHasDifferentGeneration() {
-	// Existing cacheHandle.
-	oldCacheHandle, err := chrT.cacheHandler.GetCacheHandle(chrT.object, chrT.bucket, false, 0)
+func (chrT *cacheHandlerTest) Test_addFileInfoEntryAndCreateDownloadJob_WhenJobHasCompleted() {
+	existingJob := chrT.getDownloadJobForTestObject()
+	// Make the job completed, so it's removed from job manager.
+	jobStatus, err := existingJob.Download(context.Background(), int64(chrT.object.Size), true)
+	AssertEq(nil, err)
+	AssertEq(jobStatus.Offset, chrT.object.Size)
+	// Give time for execution of callback to remove from job manager
+	time.Sleep(time.Second)
+	actualJob := chrT.jobManager.GetJob(chrT.object.Name, chrT.bucket.Name())
+	ExpectEq(nil, actualJob)
+
+	err = chrT.cacheHandler.addFileInfoEntryAndCreateDownloadJob(chrT.object, chrT.bucket)
+
+	ExpectEq(nil, err)
+	ExpectTrue(chrT.isEntryInFileInfoCache(chrT.object.Name, chrT.bucket.Name()))
+	// No new job should be added to job manager
+	actualJob = chrT.jobManager.GetJob(chrT.object.Name, chrT.bucket.Name())
+	ExpectEq(nil, actualJob)
+}
+
+func (chrT *cacheHandlerTest) Test_addFileInfoEntryAndCreateDownloadJob_WhenJobIsInvalidatedAndRemoved() {
```

Comment:
Can we rename to - Test_..._WhenUnfinishedJobIsInvalidatedAndRemoved?

If we call InvalidateAndRemove on finished job, in that case nothing will change. right?

Reply:
I think it's fine to keep it WhenJobIsInvalidatedAndRemoved because addFileInfoEntryAndCreateDownloadJob doesn't care about whether the job was unfinished or not before it was invalidated. The purpose of this test is to check the behavior of addFileInfoEntryAndCreateDownloadJob when the Job is invalidated and/or removed.

Reply:
Okay, make sense.